{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Basics of Bash for Bioinformatics\n",
    "\n",
    "by Katharina J. Hoff, giving credit to materials provided in German by Maike Tech (http://gobics.de/tech/linuxskript.pdf).\n",
    "\n",
    "It is sometimes difficult for beginners to see why they should learn all these Bash commands. References to practical application in bioinformatics are therefore denoted with a ★.\n",
    "\n",
    "## Introduction\n",
    "\n",
    "★ The shell (command line interpreter) is a powerful tool for efficiently solving many bioinformatics problems, in particular for handling large files, such as sequence files (e.g. DNA, RNA-Seq, proteins). \n",
    "\n",
    "In this course, we will use the *Bourne-again-shell*, abbreviated as *Bash*. This is only one of several available shells (another one would be the *korn-shell*, abbreviated as *ksh*). \n",
    "\n",
    "When you open Bash (and the Terminal provided by JupyterLab is running a compressed version of Bash), you see the *prompt*:\n",
    "\n",
    "\n",
    "![terminal.jpg](terminal.jpg)\n",
    "\n",
    "You enter commands after the prompt. Syntax at the beginning of the line is `username@machine-name:directory`. In above example, the username is \"I have no name!\" - which is quite unusual for a unix system. It's a phenomenon of using the JupyterLab at University of Greifswald. On you own and most other system, you'll certainly have a username that's a bit more specific. Machine name in above example is \"jupyter-hoffk83\" - the generic machine name generated by JupyterLab in Greifswald. Again: on your other systems that are not virtual machines, you'll have a more specific machine name. The directory is here abbreviated as \"~\". This symbol stands for your home directory. You type commands after the $ character. Every command ends with the Enter key from your keyboard (or with a linebreak in a script).\n",
    "\n",
    "Commands may be called with so-called *arguments*. These may be parameters that affect behavior of the program, or simple strings (strings are in the definition that we need here chains of alpha-numeric character chains generated by your keyboard). Commands and arguments are separated by a space character (the long key at the bottom of your keyboard).\n",
    "\n",
    "## Working with Bash\n",
    "\n",
    "In the following, we will look at a couple of bash commands and their usage. The first command is `echo`. This tool displays a string that was handed to `echo` as argument on the standard output device (in most cases your monitor). After execution of `echo`, we return to the shell and will be able to input further commands. Please note that the `percent-percent-script bash` is a Jupyter Notebook specific *built-in magic* command. You do not have to type that into a normal bash! If you copy the command over into the Terminal in Jupyter Lab, do not copy that part! Only copy `echo hello`. \n",
    "\n",
    "![play_button.jpb](play_button.jpg)\n",
    "\n",
    "**Hint:** for execution of command blocks (below this line, you see a command block) in Jupyter Lab, press the play-Button at the top of the notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hello\n"
     ]
    }
   ],
   "source": [
    "%%script bash \n",
    "echo hello"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "★ Even though the functionality to output strings that served as input may seem trivial, the `echo` tool is very useful in bioinformatics. For example, bioinformaticians use `echo` to automatically generate job submission scripts for compute clusters.\n",
    "\n",
    "### Parameters\n",
    "\n",
    "You can modify tool calls with parameters. These are entered after the toolname and often begin with a dash (-). For example, you can supress the trailing newline of `echo` output with parameter `-n`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hello"
     ]
    }
   ],
   "source": [
    "%%script bash \n",
    "echo -n hello"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The effect is not visible inside the Jupyter Notebook. You need to execute the command in the Terminal. The output will be\n",
    "\n",
    "![hello.jpg](hello.jpg)\n",
    "\n",
    "Have a look at this video if you don't know where to find the Terminal in JupyterHub:  https://youtu.be/HCy0VdWINpo\n",
    "\n",
    "**Syntax:**\n",
    "\n",
    "`echo` [*options*] *string*\n",
    "\n",
    "`-n` supress trailing newline\n",
    "\n",
    "Many parameters in Unix are standardized. For example, `-v` stands for *verbose* in many (not all) tools. *Verbose* means that much supplementary information is printed in additional to the usual output.\n",
    "\n",
    "### Documentation\n",
    "\n",
    "Most Unix systems have an extensive built-in documentation of Bash commands. Every tool as a *man-page*. \n",
    "\n",
    "Example:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script bash\n",
    "man less"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Above example calls the *man-page* of `less`. I did not choose this tools arbitrarily, here. In JupyterLab at University of Greifswald, *man-pages* are not all fully installed. `man echo` is for example not installed. The reason is that the computational resource center is trying to keep startup time of JupyterHub down to a minimum. So you cannot rely on built-in *man-pages* in this environment. The good news is: if you type `man echo` into http://www.google.de, you easily find the same kind of documentation on various websites. So please use a search engine when trying to access *man-pages* for JupyterLab at University of Greifswald.\n",
    "\n",
    "In a typical terminal (not inside the Jupyter Notebook output), you can navigate within the *man-page* (try in the JupyterLab terminal with `man less`): use arrow keys to go down and up, search forward with `/`+ your search term, exit with typing `q`.\n",
    "\n",
    "In general, *man-pages* consist of:\n",
    "\n",
    "* `NAME` - the tool name and a short description\n",
    "* `SYNOPSIS` - syntax of usage\n",
    "* `DESCRIPTION` - detailed description of tool\n",
    "* `OPTIONS` - description of parameters\n",
    "* `FILES` - files required by tool\n",
    "* `SEE ALSO` - hints about related tools\n",
    "* `DIAGNOSIS` - error code description\n",
    "* `BUGS` - known problems\n",
    "* `EXAMPLE` - usage example\n",
    "\n",
    "In addition, many *man-pages* provide information about `ENVIRONMENT`, `COMMANDS`, `VERSION`, `COPYRIGHT`, and `AUTHOR`. \n",
    "\n",
    "#### Formatting conventions in *man-pages*\n",
    "\n",
    "* **`bold text`** - argument has to be written exactly as stated\n",
    "* *`italic text that is underlined`* - replace the underlined text by appropriate argument\n",
    "* `[-abc]` - arguments in edgy brackets are optional\n",
    "* `-a|-b` - you can only use argument a or argument b, not both at the same time\n",
    "* `underlined argument` - the argument can be used multiple times (e.g. files)\n",
    "* `[underlined expression]` - the expression in edgy brackets can be repeated\n",
    "\n",
    "We are looking at *man-pages* from two points of views in this course:\n",
    "\n",
    "1. for retrieving information about usage of tools (because most people do not want and do not have to memorize all these parameters and options),\n",
    "2. ★ for writing our own usage information in scripts in Python, later.\n",
    "\n",
    "### Paths\n",
    "\n",
    "Unix file systems are tree-like. They begin at the root and branch out. The following is only a schematic drawing of an actual file tree.\n",
    "\n",
    "![tree.jpg](tree.jpg)\n",
    "\n",
    "When typing unix paths, always remember\n",
    "\n",
    "* the `/` is the directory and file separating character\n",
    "* **do not insert spaces** into file names, directory names for when building paths from both (I hear you: \"But spaces in directory names on xyz system work fine!\" - No, it's not fine! There are many tools that cannot deal with such spaces. Never use them. On no system! Not in file names or directory names! That's bad practice!)\n",
    "\n",
    "If you put a USB stick onto your computer in a Unix system, it will not end up in a partion comparable to Windows (e.g. `E:\\USB`). Instead, it will be mounted somewhere in the directory tree. In above drawing, you find a USB-Stick (just as an example) in `/mnt/USB`.\n",
    "\n",
    "★ In bioinformatics, it is often important to know where you are on your system. `pwd` will tell you exactly this information. It stands for *print working directory*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jovyan/Bash_basics_part_1\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We are currently in the subdirectory `jovyan/` of directory `/home/`. (The JupyterLab here returns a funny home directory... because you are a user called jovyan in the emulated machine that hosts JupyterLab. Usually, you would expect something like `/home/hoffk83` if your user name as `hoffk83`.)\n",
    "\n",
    "Every directory and every file has a unique path. (Again: **no spaces in path and file names!**) File names can be up to 255 characters. Avoid special characters, except for _ (*underscore character*). Be aware that *capitalization* matters! A capital `A` is different from a small `a`.\n",
    "\n",
    "Paths can be specified *absolute*, e.g. from the root, or *relative*. \n",
    "\n",
    "**Example:**\n",
    "\n",
    "* `/home/jovyan/Documents` - absolute path\n",
    "\n",
    "* `~/Documents` - same absolute path using the abbreviation for `~` for `/home/jovyan`.\n",
    "\n",
    "* Now, assume that we are located in `/home/jovyan/Downloads`, and we want to use a relative expression to specifiy `/home/jovyan/Documents`, without writing that absolute path. We can then say `../Documents`. The `..` mean: one directory up. (`.` means in the current directory.)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Practice:**\n",
    "\n",
    "Please test the commands given in code blocks in the Terminal of JupyterLab, as well, since the goal is not only to become fluent in a JupyterNotebook, but also to acquire knowledge in using any Bash terminal."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Management in Bash\n",
    "\n",
    "Bash is a very powerful tool for efficient data management. This makes Bash into a key environment for handling large bioinformatics files.\n",
    "\n",
    "### Files and Directories\n",
    "\n",
    "Inspect files and directory contents with `ls`. This is without doubt one of the most essential commands for checking where you are on your system, and what files are where you are.\n",
    "\n",
    "**Syntax:**\n",
    "\n",
    "`ls [Options] [Path]`\n",
    "\n",
    "**Example:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hello.jpg\n",
      "introduction_to_bash.ipynb\n",
      "play_button.jpg\n",
      "terminal.jpg\n",
      "tree.jpg\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The parameter `-l` (*long*) will show you a lot of additional information for files and directories:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 712\n",
      "-rw-r--r-- 1 38458 users  35111 Jun 24 10:42 hello.jpg\n",
      "-rw-r--r-- 1 38458 users  12745 Jun 24 10:44 introduction_to_bash.ipynb\n",
      "-rw-r--r-- 1 38458 users   9008 Jun 24 10:42 play_button.jpg\n",
      "-rw-r--r-- 1 38458 users  16164 Jun 24 10:42 terminal.jpg\n",
      "-rw-r--r-- 1 38458 users 643642 Jun 24 10:42 tree.jpg\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "ls -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Column content explained:\n",
    "    \n",
    "1. File attributes. The general format is `-rwxrwxrwx`, i.e. there are 10 positions that can be filled with some character. The first position is a `-` for files, and a `d` for directories. It may also be an `l` for link, or `p` for pipe, or `b` for block-oriented device, or `c` for character-oriented device, or `s` for socket. But `-` and `d` are the most important characters for you in this course. The next three characters show permissions for this file or directory for the owner of the file, also referred to as the `user`. `r` means read, `w` means write, and `x` stands for execute. If the characters say `rwx`, user is allows to read, write, and execute a file. If it says `---`, user is not allowed to do anything with the file. If it says e.g. `rw-`, the file can be read and written by user, but execution is forbidden. These three characters for user permissions are highly important in this course! If the execution permission is missing, an otherwise executable file will be executed. Therefore, you will later learn in this course how to modify these permissions when required. Moving to the next three characters: these are permissions for the group that owns the file. In this JupyterHub environment, the username is a number, and the group is called `users`. We will not deal in detail with changing group permissions in this course. But imagine you share the Linux computer with your room mates and your family. You don't want your parents to be able to see embarassing fotos of the latest party in your flat. So you forbid the group `parents` to view some directory with fotos. At the same time, you want your roomies to be able to view these fotos. Therefore, you will make the directory a property of `roomies` and allow them to `r-x` the directory with fotos. Next, you have three characaters for all users on the machine, again `rwx`-style. \n",
    "\n",
    "2. Number specifies the number of subdirectories of a directory. If it is a file, the number says how many names exist for this exact file (*hard links*, try to avoid using these!).\n",
    "\n",
    "3. Name of owner of file. In JupyterLab, a number. On other systems, usually a human readable user name.\n",
    "\n",
    "4. Group to whom the file belongs. In JupyterLab, mostly `user`.\n",
    "\n",
    "5. File size, by default in bytes (try `ls -lh` to get sizes in MB, GB, TB, etc.). Careful: will not display the size of contents of a directory. For directories, only the size that the folder itself requires on the file system is displayed (e.g. 4096 bytes for all directories).\n",
    "\n",
    "6. Time of last modification of file/directory.\n",
    "\n",
    "7. File or directory name.\n",
    "\n",
    "### Moving Within Your File System\n",
    "\n",
    "*Change directory*, `cd`, allowd you to change into different directories.\n",
    "\n",
    "**Syntax:**\n",
    "\n",
    "`cd [path]`\n",
    "\n",
    "**Example:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jovyan/Bioinf_pract_lecture/lecture/01_bash\n",
      "/home/jovyan\n",
      "/home\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "pwd\n",
    "cd ~/\n",
    "pwd\n",
    "cd ..\n",
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Deleting Files\n",
    "\n",
    "Be very careful about deleting (*removing*) files from your system! In Unix, files do not go into any kind of trash that can be restored. Files that are deleted are deleted. You will not be able to restore deleted files!\n",
    "\n",
    "**Syntax:**\n",
    "\n",
    "`rm [options] file`\n",
    "\n",
    "Important parameters are `-i` for *interactive* (this will require you to manually confirm the removal of files; `-r` for *recursive* removal of directories including all their contents. (Empty directories can be deleted with `rmdir`, instead.)\n",
    "\n",
    "**Example:**\n",
    "\n",
    "We will create a new (and empty) file with `touch` for the sake of demonstrating `rm` (`echo` is only used to indicate in the output at which step you are):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**Creating new file with touch new.txt\n",
      "**Confirming that the file exists:\n",
      "-rw-r--r-- 1 38458 users 0 Jun 24 10:46 new.txt\n",
      "**Deleting the new file with rm new.txt\n",
      "**Confirming that new.txt really cannot be found in your directory, anymore:\n",
      "hello.jpg\n",
      "introduction_to_bash.ipynb\n",
      "play_button.jpg\n",
      "terminal.jpg\n",
      "tree.jpg\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "echo \"**Creating new file with touch new.txt\"\n",
    "touch new.txt\n",
    "echo \"**Confirming that the file exists:\"\n",
    "ls -l new.txt\n",
    "echo \"**Deleting the new file with rm new.txt\"\n",
    "rm new.txt\n",
    "echo \"**Confirming that new.txt really cannot be found in your directory, anymore:\"\n",
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Copying files\n",
    "\n",
    "**Syntax:**\n",
    "\n",
    "`cp [options] source target`\n",
    "\n",
    "This command copies a source file to a target file. If the target file already exists, it will be overwritten without a warning. The parameter `-i` for *interactive* works in the same way as for `rm`. `-r` allows recursive copying of directories including their content.\n",
    "\n",
    "**Example:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Inspecting which files already exist:\n",
      "total 716\n",
      "-rw-r--r-- 1 38458 users  35111 Jun 24 10:42 hello.jpg\n",
      "-rw-r--r-- 1 38458 users  19195 Jun 24 10:46 introduction_to_bash.ipynb\n",
      "-rw-r--r-- 1 38458 users   9008 Jun 24 10:42 play_button.jpg\n",
      "-rw-r--r-- 1 38458 users  16164 Jun 24 10:42 terminal.jpg\n",
      "-rw-r--r-- 1 38458 users 643642 Jun 24 10:42 tree.jpg\n",
      "Copying the current notebook file:\n",
      "Inspecting effect: copied file exists:\n",
      "total 736\n",
      "-rw-r--r-- 1 38458 users  35111 Jun 24 10:42 hello.jpg\n",
      "-rw-r--r-- 1 38458 users  19195 Jun 24 10:46 introduction_to_bash.ipynb\n",
      "-rw-r--r-- 1 38458 users  19195 Jun 24 10:47 new.ipynb\n",
      "-rw-r--r-- 1 38458 users   9008 Jun 24 10:42 play_button.jpg\n",
      "-rw-r--r-- 1 38458 users  16164 Jun 24 10:42 terminal.jpg\n",
      "-rw-r--r-- 1 38458 users 643642 Jun 24 10:42 tree.jpg\n",
      "Inspecting that the copied file as been deleted (because it makes no sense to keep the copy):\n",
      "total 716\n",
      "-rw-r--r-- 1 38458 users  35111 Jun 24 10:42 hello.jpg\n",
      "-rw-r--r-- 1 38458 users  19195 Jun 24 10:46 introduction_to_bash.ipynb\n",
      "-rw-r--r-- 1 38458 users   9008 Jun 24 10:42 play_button.jpg\n",
      "-rw-r--r-- 1 38458 users  16164 Jun 24 10:42 terminal.jpg\n",
      "-rw-r--r-- 1 38458 users 643642 Jun 24 10:42 tree.jpg\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "echo \"Inspecting which files already exist:\"\n",
    "ls -l\n",
    "echo \"Copying the current notebook file:\"\n",
    "cp introduction_to_bash.ipynb new.ipynb\n",
    "echo \"Inspecting effect: copied file exists:\"\n",
    "ls -l\n",
    "rm new.ipynb\n",
    "echo \"Inspecting that the copied file as been deleted (because it makes no sense to keep the copy):\"\n",
    "ls -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Moving Files\n",
    "\n",
    "Moving files and renaming files or directories is the same thing in Unix systems, unless you move content to a different physical location. This means: moving is super fast even for huge files if you remain on the same physical location (e.g. the same hard drive). If you move to a different location, a copy process is invoked in background, and subsequently, the original file is deleted. This may of course take quite some time if the file is large.\n",
    "\n",
    "**Syntax:**\n",
    "\n",
    "`mv [options] source target`\n",
    "\n",
    "Parameter `-i` for *interactive* mode works here as well.\n",
    "\n",
    "**Example:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "** Creating a file that we can safely move around:\n",
      "-rw-r--r-- 1 38458 users 0 Jun 24 10:47 new.txt\n",
      "** Moving the file.\n",
      "** Confirming that the file was indeed moved to new2.txt:\n",
      "-rw-r--r-- 1 38458 users 0 Jun 24 10:47 new2.txt\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "echo \"** Creating a file that we can safely move around:\"\n",
    "touch new.txt\n",
    "ls -l new.txt\n",
    "echo \"** Moving the file.\"\n",
    "mv new.txt new2.txt\n",
    "echo \"** Confirming that the file was indeed moved to new2.txt:\"\n",
    "ls -l new2.txt\n",
    "rm new2.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating Directories\n",
    "\n",
    "**Syntax:**\n",
    "\n",
    "`mkdir [options] directoryname`\n",
    "\n",
    "Be careful when creating directories in subdirectories. The full path of directories needs to exist, already. Unless you give parameter `-p`, which will also create parent directories in case they are missing.\n",
    "\n",
    "**Example:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "** Inspecting that directory does not exist, yet:\n",
      "total 716\n",
      "-rw-r--r-- 1 38458 users  35111 Jun 24 10:42 hello.jpg\n",
      "-rw-r--r-- 1 38458 users  19195 Jun 24 10:46 introduction_to_bash.ipynb\n",
      "-rw-r--r-- 1 38458 users   9008 Jun 24 10:42 play_button.jpg\n",
      "-rw-r--r-- 1 38458 users  16164 Jun 24 10:42 terminal.jpg\n",
      "-rw-r--r-- 1 38458 users 643642 Jun 24 10:42 tree.jpg\n",
      "** Creating directory new_dir.\n",
      "** Inspecting that new_dir exists, now:\n",
      "total 720\n",
      "-rw-r--r-- 1 38458 users  35111 Jun 24 10:42 hello.jpg\n",
      "-rw-r--r-- 1 38458 users  19195 Jun 24 10:46 introduction_to_bash.ipynb\n",
      "drwxr-sr-x 2 38458 users   4096 Jun 24 10:47 new_dir\n",
      "-rw-r--r-- 1 38458 users   9008 Jun 24 10:42 play_button.jpg\n",
      "-rw-r--r-- 1 38458 users  16164 Jun 24 10:42 terminal.jpg\n",
      "-rw-r--r-- 1 38458 users 643642 Jun 24 10:42 tree.jpg\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "echo \"** Inspecting that directory does not exist, yet:\"\n",
    "ls -l\n",
    "echo \"** Creating directory new_dir.\"\n",
    "mkdir new_dir\n",
    "echo \"** Inspecting that new_dir exists, now:\"\n",
    "ls -l\n",
    "rmdir new_dir"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating Links\n",
    "\n",
    "**Syntax:**\n",
    "\n",
    "`ln [options] source linkname`\n",
    "\n",
    "You all know hyperlinks in the www. Links on the file system are similar. They are aliases for files or directories. There are *hard links* (please don't use these because if you ever delete a hard link, the source will be deleted as well) and *soft links*. For creating a soft link, use option `-s`. \n",
    "\n",
    "Links are particularly useful in bioinformatics when the file comes off the sequencer with an odd and very long file name. You want to keep that original file name for documentation purposes, but you do not want to spell it out during further data analysis. Thus, you create a soft link and continue to spell the link name, which is usually much shorter and easier, since you chose that link name to be short and easy.\n",
    "\n",
    "**Example:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "** Checking that the link does not exist, yet:\n",
      "total 716\n",
      "-rw-r--r-- 1 38458 users  35111 Jun 24 10:42 hello.jpg\n",
      "-rw-r--r-- 1 38458 users  19195 Jun 24 10:46 introduction_to_bash.ipynb\n",
      "-rw-r--r-- 1 38458 users      0 Jun 24 10:48 new.txt\n",
      "-rw-r--r-- 1 38458 users   9008 Jun 24 10:42 play_button.jpg\n",
      "-rw-r--r-- 1 38458 users  16164 Jun 24 10:42 terminal.jpg\n",
      "-rw-r--r-- 1 38458 users 643642 Jun 24 10:42 tree.jpg\n",
      "** Creating link.\n",
      "** Checking that hte link exists, now:\n",
      "total 716\n",
      "-rw-r--r-- 1 38458 users  35111 Jun 24 10:42 hello.jpg\n",
      "-rw-r--r-- 1 38458 users  19195 Jun 24 10:46 introduction_to_bash.ipynb\n",
      "lrwxrwxrwx 1 38458 users      7 Jun 24 10:48 new2.txt -> new.txt\n",
      "-rw-r--r-- 1 38458 users      0 Jun 24 10:48 new.txt\n",
      "-rw-r--r-- 1 38458 users   9008 Jun 24 10:42 play_button.jpg\n",
      "-rw-r--r-- 1 38458 users  16164 Jun 24 10:42 terminal.jpg\n",
      "-rw-r--r-- 1 38458 users 643642 Jun 24 10:42 tree.jpg\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "touch new.txt\n",
    "echo \"** Checking that the link does not exist, yet:\"\n",
    "ls -l\n",
    "echo \"** Creating link.\"\n",
    "ln -s new.txt new2.txt\n",
    "echo \"** Checking that hte link exists, now:\"\n",
    "ls -l\n",
    "rm new2.txt new.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Wild Cards\n",
    "\n",
    "Certain characters allow you to address many files or directories, simultaneously. You already know one wildcard: `~` abbreviates your home directory.\n",
    "\n",
    "`?` is exactly one keyboard character\n",
    "\n",
    "`*` is one, no, or possibly many characters of any type\n",
    "\n",
    "`[list of characters]` stands for exactly one character out of the list in edgy brackets\n",
    "\n",
    "`[!list of characters]` stands for excatly one character that is not part of the list in edgy brackets\n",
    "\n",
    "**Example:**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "** Creating three files that have a similar beginning.\n",
      "** Confirming that these files exist:\n",
      "total 716\n",
      "-rw-r--r-- 1 38458 users      0 Jun 24 10:48 bla\n",
      "-rw-r--r-- 1 38458 users      0 Jun 24 10:48 bli\n",
      "-rw-r--r-- 1 38458 users      0 Jun 24 10:48 blu\n",
      "-rw-r--r-- 1 38458 users  35111 Jun 24 10:42 hello.jpg\n",
      "-rw-r--r-- 1 38458 users  19195 Jun 24 10:46 introduction_to_bash.ipynb\n",
      "-rw-r--r-- 1 38458 users   9008 Jun 24 10:42 play_button.jpg\n",
      "-rw-r--r-- 1 38458 users  16164 Jun 24 10:42 terminal.jpg\n",
      "-rw-r--r-- 1 38458 users 643642 Jun 24 10:42 tree.jpg\n",
      "** Using wildcard to remove all files that start with bl.\n",
      "** Confirming that the operation worked:\n",
      "total 716\n",
      "-rw-r--r-- 1 38458 users  35111 Jun 24 10:42 hello.jpg\n",
      "-rw-r--r-- 1 38458 users  19195 Jun 24 10:46 introduction_to_bash.ipynb\n",
      "-rw-r--r-- 1 38458 users   9008 Jun 24 10:42 play_button.jpg\n",
      "-rw-r--r-- 1 38458 users  16164 Jun 24 10:42 terminal.jpg\n",
      "-rw-r--r-- 1 38458 users 643642 Jun 24 10:42 tree.jpg\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "echo \"** Creating three files that have a similar beginning.\"\n",
    "touch bla blu bli\n",
    "echo \"** Confirming that these files exist:\"\n",
    "ls -l\n",
    "echo \"** Using wildcard to remove all files that start with bl.\"\n",
    "rm bl*\n",
    "echo \"** Confirming that the operation worked:\"\n",
    "ls -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Quoting\n",
    "\n",
    "Bash automatically interprets certain special characters, so-called meta characters, with particular meanings. Sometimes, you don't want Bash to do this, but you want to address the actual character instead of the meaning. The mechanisms for this is referred to as *quoting*. For example, you might not want to refer to your home directory with `~`, but you want to refer to the character `~`. You can e.g. use a backslash for the latter, i.e. `\\~` (this is the character `~`), instead of `~` (this is the home directory).\n",
    "\n",
    "Meta characters:\n",
    "\n",
    "`; $ & ( ) < > [ ] { } ? * \" ' ~ # \\`\n",
    "\n",
    "The *backslash* protects the character that follows a backslash from interpretation by Bash. This is convenient to protect single characters, but not very conventient to protect longer character chains.\n",
    "\n",
    "*Exclamation marks* behave differently in the context of quoting, but are more convenient for protecting longer character chains.\n",
    "\n",
    "**Example:**\n",
    "\n",
    "We will use an existing environment variable `$PWD` that stores your current working directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jovyan/Bioinf_pract_lecture/lecture/01_bash\n",
      "$PWD\n",
      "/home/jovyan/Bioinf_pract_lecture/lecture/01_bash\n",
      "$PWD\n",
      "$PWD\n",
      "something /home/jovyan/Bioinf_pract_lecture/lecture/01_bash\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "echo $PWD\n",
    "echo \\$PWD\n",
    "echo \"$PWD\"\n",
    "echo \"\\$PWD\"\n",
    "echo '$PWD'\n",
    "echo 'something' $PWD"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Viewing File Contents\n",
    "\n",
    "★ In bioinformatics, we are often dealing with huge files. Opening such files with a graphic user interface editor (e.g. Notepad on Windows, or gedit or pluma on Linux) will take very much time and sometimes even fail because the file content might exceed the available memory of the computer. Nevertheless, we need to look at file contents. Unix provides a number of solutions for this.\n",
    "\n",
    "**Hint:** What happens if you try to put more into the memory of the computer than physically possible on Linux? Linux has the physical RAM, and in addition, a so-called SWAP partition. If something just *resides* in RAM and is not really needed at the moment, it is written onto the SWAP space. The SWAP partition is simply space on your harddrive, which is usually either HDD or SSD. In both cases, access and writing time is substantially slower than the access and writing time of RAM (your computer's memory). If you fill up the RAM, Linux begins to write everything that exceeds the RAM into SWAP. First of all, this slows down your system. Processes that actually require fast read/write access are now constricted by access times of your harddrive. The system might even appear to be frozen (even though it probably isn't, unless you also exceeded the SWAP space). Second, reading and writing a lot on your harddrive, in particular on a HDD, generates heat. Your computer should be able to cope with that. But let this be a warning: it has happened that computer burn because the fan was unable to cool, sufficiently. Not only on personal computers. One can fry a cluster node as well. Sad stories... so **never exceed the RAM of your machine!**\n",
    "\n",
    "#### Interactive File Viewing\n",
    "\n",
    "`less` offers the possibility to quickly view complete files. (Since you can view the entire file, loading a hugs file into `less` also may take quite some time. But it's comparably fast.)\n",
    "\n",
    "**Example:**\n",
    "\n",
    "Go to your terminal (do not try this is a JupyterNotebook code block!) and enter:\n",
    "\n",
    "`less Bash_basics_part_3/Bash_basics_part_3.ipynb`\n",
    "\n",
    "This will open your JupyterNotebook file as a plain text file, i.e. you see the source code in `less`. \n",
    "\n",
    "Navigation in `less`:\n",
    "\n",
    " * arrow keys up and down - move in document line by line\n",
    " \n",
    " * `space` or `z` - move to the next page\n",
    " \n",
    " * `q` - exit\n",
    " \n",
    " * `/` - forward search\n",
    " \n",
    " * `?` - backward search\n",
    " \n",
    "★ From my point of view as bioinformatician, `less` is one of the most frequently used Bash tools.\n",
    "\n",
    "#### Viewing the Beginning of a File\n",
    "\n",
    "Often, seeing the beginning of a file instead of its complete contents is sufficient in order to see whether the file contents adhere to a certain format. This will be very fast, even for huge files.\n",
    "\n",
    "★ In bioinformatics, the head of a file will often give us information about whether column headers are present or not, and if they are present, what columns in the file mean.\n",
    "\n",
    "**Syntax:**\n",
    "\n",
    "`head [options] file`\n",
    "\n",
    "**Example:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "** Checking that the file exists:\n",
      "-rw-r--r-- 1 38458 users 30930 Jun 24 10:48 introduction_to_bash.ipynb\n",
      "** Displaying head of file (by default 10 lines):\n",
      "{\n",
      " \"cells\": [\n",
      "  {\n",
      "   \"cell_type\": \"markdown\",\n",
      "   \"metadata\": {},\n",
      "   \"source\": [\n",
      "    \"# Basics of Bash for Bioinformatics\\n\",\n",
      "    \"\\n\",\n",
      "    \"by Katharina J. Hoff, giving credit to materials provided in German by Maike Tech (http://gobics.de/tech/linuxskript.pdf).\\n\",\n",
      "    \"\\n\",\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "echo \"** Checking that the file exists:\"\n",
    "ls -l introduction_to_bash.ipynb\n",
    "echo \"** Displaying head of file (by default 10 lines):\"\n",
    "head introduction_to_bash.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The parameter `-n` is very useful to specify the number of lines that you want to see.\n",
    "\n",
    "**Example:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "** Displaying only the first 3 lines of head of file:\n",
      "{\n",
      " \"cells\": [\n",
      "  {\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "echo \"** Displaying only the first 3 lines of head of file:\"\n",
    "head -n3 introduction_to_bash.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Viewing the End of a File\n",
    "\n",
    "Sometimes, also seeing the ending of a file instead of its complete contents is sufficient in order to see whether the file contents adhere to a certain format. This will be very fast, even for huge files.\n",
    "\n",
    "★ In bioinformatics, tools sometimes print the commands that were used to produce the file at the bottom of a file. `tail` helps us to figure out whether that's the case for a specific file.\n",
    "\n",
    "**Syntax:**\n",
    "\n",
    "`tail [options] file`\n",
    "\n",
    "**Example:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "** Displaying only the tail of file (by default 10 lines):\n",
      "   \"mimetype\": \"text/x-python\",\n",
      "   \"name\": \"python\",\n",
      "   \"nbconvert_exporter\": \"python\",\n",
      "   \"pygments_lexer\": \"ipython3\",\n",
      "   \"version\": \"3.7.6\"\n",
      "  }\n",
      " },\n",
      " \"nbformat\": 4,\n",
      " \"nbformat_minor\": 4\n",
      "}\n",
      "** Parameter -n works here, too:\n",
      " \"nbformat_minor\": 4\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "echo \"** Displaying only the tail of file (by default 10 lines):\"\n",
    "tail introduction_to_bash.ipynb\n",
    "echo \"** Parameter -n works here, too:\"\n",
    "tail -n2 introduction_to_bash.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Concatenating Files\n",
    "\n",
    "`cat` is a tool that by default will print all contents of one or several files to STDOUT. STDOUT is often your terminal window. (Hint: try the `cat` example in the terminal instead of JupyterNotebook...)\n",
    "We can redirect contents from STDOUT to files. \n",
    "\n",
    "★ In bioinformatics, `cat` can aid us in combining many files into one big file through concatenation. For example, we can concatenate different files with genomic sequence reads into one large file that will be required to execute some software, e.g. genome assembly software.\n",
    "\n",
    "**Syntax:**\n",
    "\n",
    "`cat file1 file2 ...`\n",
    "\n",
    "**Example:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "** Checking file size of original:\n",
      "-rw-r--r-- 1 38458 users 37252 Jun 24 10:51 introduction_to_bash.ipynb\n",
      "** Concatenating the file twice into a different file:\n",
      "** Checking that the output exists and is twice as large as a single input file:\n",
      "-rw-r--r-- 1 38458 users 74504 Jun 24 10:53 test.out\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "echo \"** Checking file size of original:\"\n",
    "ls -l introduction_to_bash.ipynb\n",
    "echo \"** Concatenating the file twice into a different file:\"\n",
    "cat introduction_to_bash.ipynb introduction_to_bash.ipynb > test.out\n",
    "echo \"** Checking that the output exists and is twice as large as a single input file:\"\n",
    "ls -l test.out\n",
    "rm test.out"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Counting words\n",
    "\n",
    "`wc` (word count) allows us to count the number of lines, words and bytes in a file.\n",
    "\n",
    "★ In bioinformatics, `wc` is often used to count number of lines of large sequence files. There is a useful option `-l` for this purpose!\n",
    "\n",
    "**Syntax:**\n",
    "\n",
    "`wc [options] file`\n",
    "\n",
    "**Example:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "** Checking that the file exists:\n",
      "-rw-r--r-- 1 38458 users 39934 Jun 24 10:53 introduction_to_bash.ipynb\n",
      "** Counting lines, words, bytes:\n",
      "  969  5461 39934 introduction_to_bash.ipynb\n",
      "** Counting only lines:\n",
      "969 introduction_to_bash.ipynb\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "echo \"** Checking that the file exists:\"\n",
    "ls -l introduction_to_bash.ipynb\n",
    "echo \"** Counting lines, words, bytes:\"\n",
    "wc introduction_to_bash.ipynb\n",
    "echo \"** Counting only lines:\"\n",
    "wc -l introduction_to_bash.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Handling Limited Resources of your Computer\n",
    "\n",
    "The hardware of your computer - or of any kind of cloud computing facility - has hardware limits: Harddrives/file servers have a certain size, RAM has a certain size. Be careful not to exceed both resources:\n",
    "\n",
    " \n",
    " * Harddrive of your operating system is maxed out: all the small files that are required for running your operating system cannot be executed, anymore. Your system gets \"stuck\" and cannot easily be rebooted once the harddrive is full. (If it's your own machine: keep OS partition and data partitions as separate as possible.)\n",
    "\n",
    " * RAM is maxed out (Unix): things that should go into fast read/write access are read from/written to harddrive, e.g. an HDD. This is much slower, and it produces high temperatures. If the fan of your machine fails you, you might even set the machine on fire!\n",
    "\n",
    "★ In Bioinformatics, we often have to processlarge files that produce large output files. Many tools require large RAM (e.g. assembly tools). Knowing how to monitor your system and how to prevent serious problems is therefore essential.\n",
    "\n",
    "#### Displaying Available Harddrive Space\n",
    "\n",
    " 1. `du` shows the size of a *directory*,\n",
    " 2. `df` shows the available space on all mounted storage units.\n",
    " \n",
    "Both tools have an option `-h` that is very useful for seeing large files in a human readable format instead of bytes.\n",
    " \n",
    "**Example (please test this your terminal, it takes a long time in the notebook):**\n",
    "\n",
    "![du.jpg](du.jpg)\n",
    "\n",
    "![df.jpg](df.jpg)\n",
    "\n",
    "In above example, the storage capacity of the drive that we are commonly writing into on this machine is shown in the line that contains `/home/jovyan/`.\n",
    "\n",
    "#### Minimizing Used Harddrive Space\n",
    "\n",
    "★ In Bioinformatics, it is often key to store only those original files that are required to reproduce results of a certain analysis pipeline.\n",
    "\n",
    "Many file compression standards exist. Among Windows users, `zip` is very popular. On Unix, `gzip` is commonly used. `gzip` will allow you to reduce file size of text files to approximately 60% of the original file. A nice thing about gzipped files is that you can access the contents without actually extracting the file (e.g. there's a tool `zcat` that works the same way as `cat`, except that it is made for displaying files that have been gzipped, instead of plain text files. You can extract gzipped files with `gunzip`.\n",
    "\n",
    "**Syntax:**\n",
    "\n",
    "`gzip [options] file`\n",
    "\n",
    "`gunzip [options] file`\n",
    "\n",
    "**Example:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "** Checking that the file exists:\n",
      "-rw-r--r-- 1 38458 users 44K Jun 24 10:55 introduction_to_bash.ipynb\n",
      "** Making a copy of the file because we do not want to compress our active notebook (copy has the same size as original).\n",
      "** Compressing the copied file.\n",
      "** Confirm that size has been reduced compared to original size:\n",
      "-rw-r--r-- 1 38458 users 44K Jun 24 10:55 introduction_to_bash.ipynb\n",
      "-rw-r--r-- 1 38458 users 12K Jun 24 10:56 test.ipynb.gz\n",
      "** Exracting the file (while keeping the compressed version):\n",
      "** Cleaning up.\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "echo \"** Checking that the file exists:\"\n",
    "ls -lh introduction_to_bash.ipynb\n",
    "echo \"** Making a copy of the file because we do not want to compress our active notebook (copy has the same size as original).\"\n",
    "cp introduction_to_bash.ipynb test.ipynb\n",
    "echo \"** Compressing the copied file.\"\n",
    "gzip test.ipynb\n",
    "echo \"** Confirm that size has been reduced compared to original size:\"\n",
    "ls -lh test.ipynb.gz introduction_to_bash.ipynb\n",
    "echo \"** Exracting the file (while keeping the compressed version):\"\n",
    "gunzip test.ipynb.gz\n",
    "echo \"** Cleaning up.\"\n",
    "rm test.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Often, we don't want to compress a single file, but rather a collection of files. For this, Unix has `tar` (*tape archiver*). \n",
    "\n",
    "**Syntax:**\n",
    "\n",
    "`tar -cf archive directory` for packing an archive,\n",
    "\n",
    "`tar -xf archive` for unpacking an archive.\n",
    "\n",
    "The file ending of a tarball is `.tar`.\n",
    "\n",
    "Natively, `tar` does not compress the archive unless you specify that you wish that (option `-z` for *zip*, which truly results in gzipped files/unpacking gzipped files):\n",
    "\n",
    "`tar -czf compressed_archive directory` for packing an archive,\n",
    "\n",
    "`tar -xzf compressed_archive` for unpacking an archive.\n",
    "\n",
    "The file ending of compressed tarballs is either `.tar.gz` or `.tgz` (which is an abbreviation for `tar.gz`).\n",
    "\n",
    "**Example:**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "** Changing one directory upwards.\n",
      "** Confirming that directory 01_bash exists:\n",
      "01_bash\n",
      "02_introduction_to_python\n",
      "03_variables\n",
      "04_programming_essentials\n",
      "** Making a recursive copy of the directory file because we do not want to pack our active notebook (copy has the same size as original).\n",
      "** Packing the directory including contents into a tarball (the original directory remains, which is often very handy).\n",
      "** Checking size of resulting tarball:\n",
      "-rw-r--r-- 1 38458 users 1.2M Jun 24 10:57 test.tar\n",
      "** Removing the original directory for demo purposes (if it remained, you could not see that unpacking was successful).\n",
      "total 1.2M\n",
      "drwxr-sr-x 3 38458 users 4.0K Jun 24 10:57 01_bash\n",
      "drwxr-sr-x 2 38458 users 4.0K Jun 24 08:49 02_introduction_to_python\n",
      "drwxr-sr-x 3 38458 users 4.0K Jun 24 09:03 03_variables\n",
      "drwxr-sr-x 3 38458 users 4.0K Jun 24 09:05 04_programming_essentials\n",
      "-rw-r--r-- 1 38458 users 1.2M Jun 24 10:57 test.tar\n",
      "** Unpacking the uncompressed tarball:\n",
      "** Confirming that unpacking has happend:\n",
      "total 1.2M\n",
      "drwxr-sr-x 3 38458 users 4.0K Jun 24 10:57 01_bash\n",
      "drwxr-sr-x 2 38458 users 4.0K Jun 24 08:49 02_introduction_to_python\n",
      "drwxr-sr-x 3 38458 users 4.0K Jun 24 09:03 03_variables\n",
      "drwxr-sr-x 3 38458 users 4.0K Jun 24 09:05 04_programming_essentials\n",
      "drwxr-sr-x 3 38458 users 4.0K Jun 24 10:57 test\n",
      "-rw-r--r-- 1 38458 users 1.2M Jun 24 10:57 test.tar\n",
      "** Now creating a compressed tarball.\n",
      "** Now unpacking a compressed tarball.\n",
      "** Cleaning up.\n"
     ]
    }
   ],
   "source": [
    "%%script bash\n",
    "echo \"** Changing one directory upwards.\"\n",
    "cd ../\n",
    "echo \"** Confirming that directory 01_bash exists:\"\n",
    "ls\n",
    "echo \"** Making a recursive copy of the directory file because we do not want to pack our active notebook (copy has the same size as original).\"\n",
    "cp -r 01_bash test\n",
    "echo \"** Packing the directory including contents into a tarball (the original directory remains, which is often very handy).\"\n",
    "tar -cf test.tar test\n",
    "echo \"** Checking size of resulting tarball:\"\n",
    "ls -lh test.tar\n",
    "echo \"** Removing the original directory for demo purposes (if it remained, you could not see that unpacking was successful).\"\n",
    "rm -r test\n",
    "ls -lh\n",
    "echo \"** Unpacking the uncompressed tarball:\"\n",
    "tar -xf test.tar\n",
    "echo \"** Confirming that unpacking has happend:\"\n",
    "ls -lh\n",
    "echo \"** Now creating a compressed tarball.\"\n",
    "tar -czf test.tar.gz test\n",
    "echo \"** Now unpacking a compressed tarball.\"\n",
    "tar -xzf test.tar.gz\n",
    "echo \"** Cleaning up.\"\n",
    "rm -r test test.tar test.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you receive a compressed file that had been created on Windows, it's often a `zip` file. You can easily extract such files on Unix systems with the command `unzip file.zip`.\n",
    "\n",
    "#### Process Management\n",
    "\n",
    "Every process that is running on your Unix system has an ID, the *Process ID*, in short PID. Knowing the PID of your processes that are important to you is very helpful because with that ID, you can for example kill your own processes that have gone wild (e.g. in terms of RAM consumption or in terms of filling your hard drives).\n",
    "\n",
    "##### Foreground Processes\n",
    "\n",
    "We call a process a *foreground* process if it is running in your terminal, and the terminal is in the meanwile blocked by that process.\n",
    "\n",
    "**Example (hold off trying this in your terminal until you have learned how to kill processes!):**\n",
    "\n",
    "![foreground.jpg](foreground.jpg)\n",
    "\n",
    "In this example, a never ending loop is occupying the terminal window. No other commands can be entered. (You can kill the process with `CTRL + c`.)\n",
    "\n",
    "##### Background Processes\n",
    "\n",
    "*Background* processes are started in your terminal and continue running - but you are free to enter other commands into your terminal, in the meantime. In order to send you process to background, append the operator `&` before hitting Enter. You will then also see the PID of your process that was sent into background. Remember such PIDs in order to be able to kill the process if required. Important: a background process remains attached to your terminal! This means: if you close a terminal, the background process automatically dies. If you later need to be able to work with processes independent of particular terminal sessions, please check out `nohup` for decoupling your process from the current terminal. Also, read about `screen`, which will not decouple your process from the terminal, but keep terminal sessions open even if you loose your network connection to the machine that is running your terminal.\n",
    "\n",
    "\n",
    "**Example (hold off trying this in your terminal until you have learned how to kill processes!):**\n",
    "\n",
    "![background.jpg](background.jpg)\n",
    "\n",
    "In above example, the loop that prints \"Hello world\" is sent to background, and we are able to see the PID 147. (This command will continue to print into STDOUT, i.e. your terminal, but you can always press enter and return to a clear command prompt that will enable you to start additional commands while the loop still runs in background.)\n",
    "\n",
    "If you forgot to send you process to background, you can send an already running foreground process to background as follows:\n",
    "\n",
    " 1.) start the process as foreground process\n",
    " \n",
    " 2.) press `CTRL + z` to put the process on hold\n",
    " \n",
    " 3.) type `bg` into your terminal to send the process to background (it will resume there)\n",
    " \n",
    "You can retrieve processes from background into foreground by typing `fg`.\n",
    "\n",
    "**Example (hold off trying this in your terminal until you have learned how to kill processes!):**\n",
    "\n",
    "![fg.jpg](fg.jpg)\n",
    "\n",
    "#### Ending Processes\n",
    "\n",
    "One option to kill a process is by pressing `CTRL + c`. If the process is *friendly*, that will kill a foreground process. This approach may also fail (e.g. if you have maxed your RAM).\n",
    "\n",
    "The command `kill` offers you the option to kill processes by their PID.\n",
    "\n",
    "**Syntax:**\n",
    "\n",
    "`kill [signal] PID`\n",
    "\n",
    "**Example:**\n",
    "\n",
    "![kill.jpg](kill.jpg)\n",
    "\n",
    "When testing this example yourself, you will of course have to use the individual PID of your job, and not the number 147.\n",
    "\n",
    "The standard signal of `kill` is `-15` (i.e. it is executed with `-15`, even if you don't specify a signal). This signal asks the process kindly whether it wants to kill itself. The process will often do you that favor. Similar to pressing `CTRL + c`. In works in many cases.\n",
    "\n",
    "Sometimes, a default `kill` will fail to end your process. For such cases, the signal **`-9`** is very useful because it withdraws all system resources from the process. The process is forced to die. Keep this in mind!\n",
    "\n",
    "**Example:**\n",
    "\n",
    "`kill -9 147`\n",
    "\n",
    "#### Watching Processes and Monitoring Resources\n",
    "\n",
    "The command `ps` (*process status*) shows current processes, the status of process and some more information.\n",
    "\n",
    "**Example:**\n",
    "\n",
    "![ps.jpg](ps.jpg)\n",
    "\n",
    "Columns:\n",
    "\n",
    " * `PID` the process ID\n",
    " \n",
    " * `TTY` *tele type* shows, which terminal started the process (you may have several terminals running simultaneously, also in JupyterHub)\n",
    " \n",
    " * `TIME` time that has elapsed since the process started\n",
    " \n",
    " * `CMD` the command that started this process (often a tool name)\n",
    " \n",
    "The tool `top` shows running processes *and their resources consumption*. `top` will run as a foreground process. You can exit the `top` table display by pressing `q`.\n",
    "\n",
    "★ In bioinformatics, `top` is a highly valuable tool for monitoring jobs that process large amounts of data. For example, the output table can give you the PID of a job that you are looking for, you can easily see whether RAM is close to being exceeded, you can also easily see whether CPU is used efficiently.\n",
    "\n",
    "Start by typing\n",
    "\n",
    "`top`\n",
    "\n",
    "in your terminal.\n",
    "\n",
    "![top.jpg](top.jpg)\n",
    "\n",
    "In above image, we seed the following columns:\n",
    "\n",
    " * `PID` - process ID\n",
    " \n",
    " * `USER` - owner of a process\n",
    " \n",
    " * `PR` - priority of a process\n",
    " \n",
    " * `NI` - nice value of the process (this value changes job priority)\n",
    " \n",
    " * `S` - status of process (`D` is an uninterruptible sleep, `R` is running, `S` is sleeping, `T` is traced or stopped, `Z` is a zombie)\n",
    " \n",
    " * `%CPU` - proportion of CPU usage (in %) that a job occupies; be aware that if you have more than one core, the percentage may exceed 100% (e.g. 4 cores -> increase up to 400%)\n",
    " \n",
    " * `%MEM` - proportion of RAM (in %) that a process occupies\n",
    " \n",
    " * `TIME+` - runtime of the job until now\n",
    " \n",
    " * `COMMAND` - name of the command/tool\n",
    " \n",
    "The process priority can be changed. This is particularly important to know if you later work on a multi-user system. It's less important in the JupyterHub environment where you are typically the only user. When needed, please inform yourself about the tool `nice`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
